### PR TITLE
Add 'Skip evaluation' button to quick review mode

### PR DESCRIPTION
**Pull request by rgwohlbold**
This pull request adds a "Skip evaluation button" to the quick review mode, see #1508 .

### CODE REVIEW

File: evap/staff/templates/quick-review.js [hunk 1]
```diff
@@ -111,6 +113,35 @@ $(document).ready(() => {
         }
     }
 
+    var nextEvaluationIndex = 0;
+    setNextEvaluation(0)
```
Comment (janno42) [inline evap/staff/templates/quick-review.js:117]: there is a semicolon missing here :)

File: evap/staff/templates/quick-review.js [hunk 2]
```diff
@@ -111,6 +113,35 @@ $(document).ready(() => {
...
+    $("#skip-next-evaluation").on("click", event => {
+        const skippedEvaluationId = $("[data-evaluation").not(":hidden").data("evaluation")
+        $.ajax({
+            type: "POST",
+            data: {evaluation_id: skippedEvaluationId},
+            url: "{% url 'staff:evaluation_textanswers_skip' %}",
+            error: function(){ window.alert("{% trans 'The server is not responding.' %}"); }
+        });
+        setNextEvaluation(++nextEvaluationIndex);
+    })
```
Comment (janno42) [inline evap/staff/templates/quick-review.js:128]: here as well

**Review by janno42 — COMMENTED** [review]
looks good!
you don't need to do translations, all strings seem to be correctly marked for translation already and i'll do the rest with the next production update.

and you are right: a test not returning a skipped evaluation on requesting `find_unreviewed_evaluations` would be nice.

File: evap/staff/templates/quick-review.js [hunk 3]
```diff
@@ -111,6 +113,35 @@ $(document).ready(() => {
         }
     }
 
+    var nextEvaluationIndex = 0;
+    setNextEvaluation(0);
+
+    $("#skip-next-evaluation").on("click", event => {
+        const skippedEvaluationId = $("[data-evaluation").not(":hidden").data("evaluation");
```
Comment (pixunil) [inline evap/staff/templates/quick-review.js:120]:
```suggestion
        const skippedEvaluationId = $("[data-evaluation]").not(":hidden").data("evaluation");
```
Unless the selector without `]` has a special meaning unknown to me

File: evap/staff/templates/staff_evaluation_textanswers_quick.html [hunk 4]
```diff
@@ -76,10 +77,12 @@ <h5 class="modal-title" id="hotkeys-modal-title">Hotkeys</h5>
                             {% trans 'There are no text answers for this evaluation.' %}
                         </span>
                     {% endif %}
-                    {% if next_evaluation %}
-                        <span>
-                            {% blocktrans with name=next_evaluation.full_name count answers=next_evaluation.num_unreviewed_textanswers %}The next evaluation "{{ name }}" has got {{ answers }} unreviewed text answer.{% plural %}The next evaluation "{{ name }}" has got {{ answers }} unreviewed text answers.{% endblocktrans %}
-                        </span>
+                    {% if next_evaluations %}
+                        {% for next_evaluation in next_evaluations %}
+                            <span data-index="{{ forloop.counter0 }}" data-evaluation="{{ next_evaluation.pk }}">
```
Comment (pixunil) [inline evap/staff/templates/staff_evaluation_textanswers_quick.html:82]: `[data-index]` is a bit too generic for my taste. Would it suffice to include `[data-evaluation]` and depending on the currently visible next evaluation to change the `[href]` of `[data-url=next-evaluation]`? Or include a `[data-next-evaluation-url]` here?

**Review by pixunil — CHANGES_REQUESTED** [review]
Nice solution :cat:

File: evap/staff/templates/staff_evaluation_textanswers_quick.html (hunk 4 above)
Comment (rgwohlbold) [reply evap/staff/templates/staff_evaluation_textanswers_quick.html:82]: I don't think this would be enough. We don't only need the url, but also the title. I will rename it to data-next-evaluation-index

**Comment by rgwohlbold** [comment]
I added a test, but since there is no way to check if a review will occur again without parsing HTML, the test only checks that the HTTP status is 200

File: evap/staff/templates/staff_evaluation_textanswers_quick.html [hunk 5]
```diff
@@ -21,6 +21,7 @@ <h5 class="modal-title" id="hotkeys-modal-title">Hotkeys</h5>
                     <tr><td><kbd>&#x21b2;</kbd></td><td>{% trans 'Review next evaluation' %}</td></tr>
                     <tr><td><kbd>N</kbd></td><td>{% trans 'Show all again' %}</td></tr>
                     <tr><td><kbd>M</kbd></td><td>{% trans 'Show unreviewed' %}</td></tr>
+                    <tr><td><kbd>S</kbd></td><td>{% trans 'Skip evaluation' %}</td></tr>
```
Comment (janno42) [inline evap/staff/templates/staff_evaluation_textanswers_quick.html:24]: please move the new entry between "Review next evaluation" and "Show all again", then the order of the hotkeys and buttons is consistent.

**Review by janno42 — CHANGES_REQUESTED** [review]
when skipping an evaluation and then pressing enter (hotkey for next evaluation), the skipped evaluation will still show up for review. it works correctly when using the button instead.

**Comment by rgwohlbold** [comment]
Implemented changes, fixed tests :+1:

**Review by janno42 — APPROVED** [review]

File: evap/staff/tests/test_views.py [hunk 6]
```diff
@@ -2383,6 +2383,30 @@ def test_finishing_review_updates_results(self):
...
+        )
+        top_general_questionnaire = baker.make(Questionnaire, type=Questionnaire.Type.TOP)
+        baker.make(Question, questionnaire=top_general_questionnaire, type=Question.TEXT)
+        evaluation.general_contribution.questionnaires.set([top_general_questionnaire])
+        baker.make(TextAnswer, contribution=evaluation.general_contribution)
+
+        skip_url = "/staff/textanswers/skip"
+        self.app.post(skip_url, user=manager, status=200, params={
+            'evaluation_id': evaluation.id
+        })
```
Comment (pixunil) [inline evap/staff/tests/test_views.py:2407]: You should test some behaviour in your test, here it would be to check that the skipped evaluation is saved in the session. This works on my machine: (why `self.client.session` is not working, but `repsonse.client.session` does – I don’t know)
```suggestion
        response = self.app.post(skip_url, user=manager, status=200, params={
            'evaluation_id': evaluation.id,
        })
        self.assertEqual(response.client.session['review-skipped'], {evaluation.id})
```

**Review by pixunil — CHANGES_REQUESTED** [review]

File: evap/staff/templates/quick-review.js [hunk 7]
```diff
@@ -22,20 +22,22 @@ $(document).ready(() => {
...
+            "j":            "[data-action=publish]",
+            "k":            "[data-action=make_private]",
+            "l":            "[data-action=hide]",
+            "backspace":    "[data-action=unreview]",
+            "enter":        `[data-url=next-evaluation][data-next-evaluation-index=${nextEvaluationIndex}]`,
+            "m":            "[data-startover=unreviewed]",
+            "n":            "[data-startover=all]",
+            "s":            "[data-skip-evaluation]",
         };
+        const action = actions[event.key.toLowerCase()]
```
[the remark this answers is not in the document]
Comment (rgwohlbold) [reply evap/staff/templates/quick-review.js:37]: I think the readability of the code should be preferred over not using the object as a map. I will add the semicolon.

File: evap/staff/templates/quick-review.js [hunk 8]
```diff
@@ -111,6 +113,37 @@ $(document).ready(() => {
         }
     }
 
+    let nextEvaluationIndex = 0;
+    updateNextEvaluation(0);
+
+    $("[data-skip-evaluation]").on("click", event => {
+        const skippedEvaluationId = $("[data-evaluation]").not(":hidden").data("evaluation");
+        if (skippedEvaluationId !== undefined) {
```
Comment (He3lixxx) [inline evap/staff/templates/quick-review.js:121]: In what case is this equal to undefined?

File: evap/staff/views.py [hunk 9]
```diff
@@ -1106,20 +1106,35 @@ def evaluation_textanswers(request, semester_id, evaluation_id):
...
 
     template_data.update(dict(evaluation_sections=evaluation_sections, contributor_sections=contributor_sections))
     return render(request, "staff_evaluation_textanswers_full.html", template_data)
 
+@reviewer_required
+def evaluation_textanswers_skip(request):
+    evaluation_id = request.POST["evaluation_id"]
+    evaluation = get_object_or_404(Evaluation, id=evaluation_id)
+    if evaluation.is_fully_reviewed:
+        raise PermissionDenied
```
Comment (He3lixxx) [inline evap/staff/views.py:1131]: Why is this required?

File: evap/staff/tests/test_views.py [hunk 10]
```diff
@@ -2383,6 +2383,31 @@ def test_finishing_review_updates_results(self):
...
+        student = baker.make(UserProfile, email="student@institution.example.com")
+        evaluation = baker.make(
+            Evaluation,
+            participants=[student],
+            voters=[student],
+            state="in_evaluation",
+            can_publish_text_results=True,
+        )
+        top_general_questionnaire = baker.make(Questionnaire, type=Questionnaire.Type.TOP)
+        baker.make(Question, questionnaire=top_general_questionnaire, type=Question.TEXT)
```
Comment (He3lixxx) [inline evap/staff/tests/test_views.py:2400]: Could you see if the test works without some of these models? (I would especially expect that these two are not necessary for the test to pass)

**Review by pixunil — APPROVED** [review]

**Review by He3lixxx — COMMENTED** [review]
Very cool, I'm impressed.

With this code, I have to say I've never seen the UI nor the whole js code here so far, so I don't feel able to properly review this.

File: evap/staff/views.py (hunk 9 above)
Comment (rgwohlbold) [reply evap/staff/views.py:1131]: I will remove this. Can I remove the two lines above as well?

File: evap/staff/tests/test_views.py (hunk 10 above)
Comment (rgwohlbold) [reply evap/staff/tests/test_views.py:2400]: When I remove the is_fully_reviewed check, this can be removed as well

File: evap/staff/templates/quick-review.js (hunk 8 above)
Comment (rgwohlbold) [reply evap/staff/templates/quick-review.js:121]: When a user presses the "skip" key, the button is pressed (virtually). This check assures that no AJAX request is sent if the "skip evaluation" button is hidden, in which case skippedEvaluationId is undefined

**Review by He3lixxx — APPROVED** [review]
